### PR TITLE
Issue #26: Remove compound primary keys on temporal clock tables

### PR DESCRIPTION
**Pull request by bijanvakili**
* Issue #26 : Replace compound primary key `(tick, entity_id)` on temporal clock tables with
    * Primary key `id` as `UUID` column
    * Unique constraint on `(tick, entity_id)`
* Bump version to 0.3.3

### CODE REVIEW

**Comment by dave-clover** [comment]
Curious about why?

**Comment by pauljz** [comment]
Django doesn't natively support compound primary keys.

File: temporal_sqlalchemy/bases.py [hunk 1]
```diff
@@ -21,9 +23,9 @@
 
 
 class EntityClock(object):
-    tick = sa.Column(sa.Integer, primary_key=True, autoincrement=False)
-    timestamp = sa.Column(sa.DateTime(True),
-                          server_default=sa.func.current_timestamp())
+    id = sa.Column(sap.UUID(as_uuid=True), default=uuid.uuid4, primary_key=True)
+    tick = sa.Column(sa.Integer, nullable=False)
+    timestamp = sa.Column(sa.DateTime(True), server_default=sa.func.current_timestamp())
```
Comment (pauljz) [inline temporal_sqlalchemy/bases.py]: Do we need to explicitly set nullable=False here too? I see we did it in the migration, but I'm not sure how sqlalchemy treats this.

File: temporal_sqlalchemy/clock.py [hunk 2]
```diff
@@ -131,24 +131,30 @@ def make_temporal(cls: nine.Type[Clocked]):
...
                 sa.ForeignKey(activity_cls.id), nullable=False)
             clock_properties['activity'] = orm.relationship(
                 activity_cls, backref=backref_name)
-            clock_properties['__table_args__'] = (
-                sa.UniqueConstraint('entity_id', 'activity_id'),
-                {'schema': schema}
-            )
-        else:
-            clock_properties['__table_args__'] = {'schema': schema}
+            table_args.append(sa.UniqueConstraint('entity_id', 'activity_id'))
```
Comment (pauljz) [inline temporal_sqlalchemy/clock.py]: The uniqueness constraint creates an implicit index, right? Wouldn't want to fall back on sequential scans here.

File: temporal_sqlalchemy/bases.py (hunk 1 above)
Comment (bijanvakili) [inline temporal_sqlalchemy/bases.py]: It's not up to SQLAlchemy.  It's an implicit requirement in Postgres.

See the Postgres's documentation for [CREATE TABLE](https://www.postgresql.org/docs/9.5/static/sql-createtable.html):

"The PRIMARY KEY constraint specifies that a column or columns of a table can contain only unique (non-duplicate), nonnull values."

File: temporal_sqlalchemy/clock.py (hunk 2 above)
Comment (bijanvakili) [inline temporal_sqlalchemy/clock.py]: Yes it creates a `btree` index:

```
\d+ trg_history.guac_prior_auth_correspondence_events_clock
...
Indexes:
    "guac_prior_auth_correspondence_events_clock_pkey" PRIMARY KEY, btree (id)
    "guac_prior_auth_correspondence_events_clock_tick_entity_id_key" UNIQUE CONSTRAINT, btree (tick, entity_id)
```

And I've verified it will use an index scan:

```
dev_app_db=# EXPLAIN SELECT * FROM trg_history.guac_prior_auth_correspondence_events_clock ORDER BY tick, entity_id;
                                                                              QUERY PLAN
----------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Index Scan using guac_prior_auth_correspondence_events_clock_tick_entity_id_key on guac_prior_auth_correspondence_events_clock  (cost=0.15..57.95 rows=920 width=60)
```

File: temporal_sqlalchemy/bases.py (hunk 1 above)
Comment (pauljz) [inline temporal_sqlalchemy/bases.py]: the timestamp isn't part of the primary key though

Comment (bijanvakili) [inline temporal_sqlalchemy/bases.py]: The `timestamp` column wasn't changed in this PR.  I only fixed a code indent.

Adding `nullable=False` here wouldn't add much value since it has a server generated default. The `entity_id` columns on the history tables are also missing a `NOT NULL` constraint.  So if you strongly believe it should be fixed, it should go into a separate PR as this one is just to remove the compound primary keys.
